prov/efa/docs/efa_rdm_protocol_v4.md

```diff
@@ -0,0 +1,1318 @@
+
+Table: 3.15 the binary format of an ATOMRSP packet header.
+
+| Name | Length (bytes) | type | C language type | Note |
+|-|-|-|-|-|
+| `type`           | 1 | integer | `uint8_t`  | part of base header |
+| `version`        | 1 | integer | `uint8_t`  | part of base header|
+| `flags`          | 2 | integer | `uint16_t` | part of base header |
+| `padding`	   | 4 | integer | `uint32_t` | alignment for 8 bytes |
+| `reserved`       | 4 | integer | `uint32_t` | reserved for future use |
+| `recv_id`        | 4 | integer | `uint32_t` | ID of the receive operation on the requester side |
+| `seg_length`     | 8 | integer | `uint64_t` | length of the application data in the packet |
+| `connid`         | 4 | integer | `uint32_t` | sender connection ID, this field is optional see section 4.4 for more details |
+
+## 4. Extra features and requests
+
+This chapter describes the extra features and requests of protocol v4.
+
+### 4.1 RDMA read based data transfer (RDMA read)
+
+The extra feature "RDMA read based data transfer" (RDMA read) was introduced together
+with protocol v4, when libfabric 1.10 was released. It was assigned ID 0.
+
+It is defined as an extra feature because there is a set of requirements (firmware,
+EFA kernel module and rdma-core) to be met before an endpoint can use the RDMA
+read capability, therefore an endpoint cannot assume the other party support RDMA read.
+
+The "RDMA read" extra feature corresponds to the following sub-protocols:
+long-read message, emulated long-read write, direct read.
+
+#### Long-read message sub-protocol
+
+The long-read message sub-protocol uses RDMA read to implement two-sided communication.
+
+The work flow of long-read message sub-protocol is illustrated in the following diagram:
+
+![long-read message](message_longread.png)
+
+There are two packet types involved in this protocol: LONGREAD_RTM and EOR (End Of Read).
+
+LONGREAD_RTM is sent by the sender to initiate the communication. The binary format of 
+a LONGREAD_RTM packet's mandatory header is listed in table 4.1
+
+Table: 4.1 the binary format of a LONGREAD_RTM packet's mandatory header
+
+| Name | Length (bytes) | type | C language type | Note |
+|-|-|-|-|-|
+| `type`           | 1 | integer | `uint8_t`  | part of base header |
+| `version`        | 1 | integer | `uint8_t`  | part of base header|
+| `flags`          | 2 | integer | `uint16_t` | part of base header |
+| `msg_id`         | 4 | integer | `uint32_t` | message ID |
+| `msg_length`     | 4 | integer | `uint64_t` | total length of the message |
+| `send_id`        | 4 | integer | `uint32_t` | ID of the receive operation  |
+| `read_iov_count` | 4 | integer | `uint32_t` | number of iov to read |
+| `read_iov`       | `read_iov_count` * 24 | array | `efa_rma_iov[]` | read iov information |
+
+In LONGREAD_RTM packet mandatory header, there are two new fields `read_iov_count` and `read_iov`.
```

Comment:
This is probably too specific for a protocol doc, instead maybe just state that read_iov needs contain the right descriptor so that RDMA read can be used to retrieve the data from the sender's buffer?